core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java

```diff
@@ -74,29 +86,47 @@ public void heartbeat() {
     }
 
     try {
-      UfsManager.UfsClient ufsclient = mMountTable.getUfsClient(mMountId);
+      UfsManager.UfsClient ufsclient = Objects.requireNonNull(mMountTable.getUfsClient(mMountId));
       try (CloseableResource<UnderFileSystem> ufsResource = ufsclient.acquireUfsResource()) {
         UnderFileSystem ufs = ufsResource.get();
         if (ufs.supportsActiveSync()) {
           SyncInfo syncInfo = ufs.getActiveSyncInfo();
           // This returns a list of ufsUris that we need to sync.
           Set<AlluxioURI> ufsSyncPoints = syncInfo.getSyncPoints();
           // Parallelize across sync points
-          List<Callable<Void>> tasksPerSyncPoint = new ArrayList<>(ufsSyncPoints.size());
+          List<CompletableFuture<Long>> tasksPerSync = new ArrayList<>();
           for (AlluxioURI ufsUri : ufsSyncPoints) {
-            tasksPerSyncPoint.add(() -> {
+            tasksPerSync.add(CompletableFuture.supplyAsync(() -> {
               processSyncPoint(ufsUri, syncInfo);
-              return null;
-            });
+              return syncInfo.getTxId();
+            }, mSyncManager.getExecutor()));
           }
-          mSyncManager.getExecutor().invokeAll(tasksPerSyncPoint);
           // Journal the latest processed txId
-          mFileSystemMaster.recordActiveSyncTxid(syncInfo.getTxId(), mMountId);
```

Comment:
Was this wrong before, journaling too early?

Reply:
no, it worked properly before because the sync occurred synchronously within the heartbeat